### PR TITLE
[network, WIP] start a Libp2p DHT + add a topology probing its peers

### PR DESCRIPTION
**Pull request by huitseeker**
Meant to inform #1040. Do not merge, for informational purposes only.

### CODE REVIEW

File: network/topology/dhtPeersTopology.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+func NewDHTPeersTopology(nodeID flow.Identity, routing *discovery.RoutingDiscovery) DHTPeersTopology {
+	return DHTPeersTopology{
+		fixedNodeID: nodeID,
+		routing:     routing,
+	}
+}
+
+func (r DHTPeersTopology) GenerateFanout(ids flow.IdentityList, _ network.ChannelList) (flow.IdentityList, error) {
```
Comment (vishalchangrani) [inline network/topology/dhtPeersTopology.go:30]: I was hoping to not use Topology and peer manager at all in the unstaked side of things. LibP2P `pubsub` accepts a `discovery` implementation and then [manages it for us](https://github.com/libp2p/go-libp2p-pubsub/blob/master/discovery.go) . So once we have our [DHT server setup](https://github.com/libp2p/go-libp2p-pubsub/blob/master/discovery.go#L8), we can pass it in as a pubsub options.

Our `Topology` and `PeerManager` we just not designed for such a thing.

Comment (huitseeker) [reply network/topology/dhtPeersTopology.go:30]: I'm happy with that. You can just cherry-pick the first commit, which just generates the discovery.

File: network/p2p/dht/dht.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		dht.RoutingTableRefreshPeriod(routingTableRefresh),        // this is 100 seconds
+		dht.RoutingTableRefreshQueryTimeout(peerDiscoveryTimeout), // this is 2 seconds
+
+		// public capabilities we don't need
+		dht.DisableProviders(),
+		dht.DisableValues(),
+	}
+
+	// If we have no bootstrapPeers, we're the server
+	if len(bootstrapPeers) == 0 {
```
Comment (smnzhu) [inline network/p2p/dht/dht.go:35]: Can you explain this? If there are no bootstrap peers, how does the node join the DHT network?

Comment (huitseeker) [reply network/p2p/dht/dht.go:35]: It is a server that expects the other nodes to join it (and accepts them as peers). That would be the behavior of the staked ANs.

In particular, a piece of code that's not in this WIP is how the unstaked ANs are started through a constructor that passes a non-empty list of bootstrap peers here (they pass in specifically the staked ANs).

Comment (smnzhu) [reply network/p2p/dht/dht.go:35]: But, the staked AN needs to know about the other staked AN's which participate in the DHT right? 

My understanding of how we were going to set this up was that the staked AN's together would form a DHT, and the values stored in this DHT would be the node information of the unstaked nodes (but perhaps this isn't accurate?).

Are we relying on the fact that the staked AN's already know about each other through bootstrap data / protocol state, and therefore don't have to explicitly connect to each other here?

Comment (huitseeker) [reply network/p2p/dht/dht.go:35]: This is a demo PR, not the actual data flow.

> My understanding of how we were going to set this up was that the staked AN's together would form a DHT, and the values stored in this DHT would be the node information of the unstaked nodes (but perhaps this isn't accurate?).

Yeah, that's accurate, but just not done in this PR, which doesn't so far change the bootstrap of the node.

**Comment by huitseeker** [comment]
Demo made obsolete by #1098

**Pull request closed without merging**